hydroflow_lang/src/graph/flat_graph.rs

```diff
@@ -322,197 +322,288 @@ impl FlatGraphBuilder {
+                    .map(|(port, _pred)| port)
+                    .cloned()
+                    .collect();
+
+                // Collect output arguments (successors).
+                let mut output_edges: Vec<(&PortIndexValue, GraphNodeId)> = self
+                    .flat_graph
+                    .graph
+                    .successors(node_id)
+                    .map(|(edge_id, succ)| (&self.flat_graph.ports[edge_id].0, succ))
+                    .collect();
+                // Ensure sorted by port index.
+                output_edges.sort();
+
+                let output_ports: Vec<PortIndexValue> = output_edges
+                    .into_iter()
+                    .map(|(port, _succ)| port)
+                    .cloned()
+                    .collect();
+
+                (input_ports, output_ports)
+            };
+
+            // Generic arguments.
+            let generics = get_operator_generics(&mut self.diagnostics, operator);
+            // Generic argument errors.
+            {
+                if !op_constraints
+                    .persistence_args
+                    .contains(&generics.persistence_args.len())
+                {
+                    self.diagnostics.push(Diagnostic::spanned(
+                        generics.generic_args.span(),
+                        Level::Error,
+                        format!(
+                            "`{}` should have {} persistence lifetime arguments, actually has {}.",
+                            op_constraints.name,
+                            op_constraints.persistence_args.human_string(),
+                            generics.persistence_args.len()
+                        ),
+                    ));
+                }
+                if !op_constraints.type_args.contains(&generics.type_args.len()) {
+                    self.diagnostics.push(Diagnostic::spanned(
+                        generics.generic_args.span(),
+                        Level::Error,
+                        format!(
+                            "`{}` should have {} generic type arguments, actually has {}.",
+                            op_constraints.name,
+                            op_constraints.type_args.human_string(),
+                            generics.type_args.len()
+                        ),
+                    ));
+                }
+            }
+
+            self.flat_graph.operator_instances.insert(
+                node_id,
+                OperatorInstance {
+                    op_constraints,
+                    input_ports,
+                    output_ports,
+                    generics,
+                    arguments: operator.args.clone(),
+                },
+            );
+        }
+    }
+
     /// Validates that operators have valid number of inputs, outputs, & arguments.
     /// Adds errors (and warnings) to `self.diagnostics`.
     /// TODO(mingwei): Clean this up, make it do more than just arity? Do no overlapping edge ports.
```

Comment:
Can you remove this TODO now? Are the other checks now handled elsewhere?

Reply:
Yeah the other checks are in `connect_operator_links` and `make_operator_instances`